### PR TITLE
Fetch quotes json

### PR DESCRIPTION
**Pull request by MatthewBerglund**
Moved quotes into quotes.json. Used Fetch API to request quotes, then select a random quote from the fetched JSON data.

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -1,60 +1,59 @@
...
-  const twitterLink = document.getElementById('tweet-quote-link');
-  const intentAddress = `https://twitter.com/intent/tweet?text=“${quote.text}”  — ${quote.movie}`;
-  twitterLink.setAttribute('href', intentAddress);
+fetch('quotes.json')
+  .then(response => {
+    if (!response.ok) {
+      document.getElementById('quote-box').textContent = `
+        Network request failed with status 
+        ${response.status}: ${response.statusText}. 
+        Please refresh to try again.
```
Comment (motine) [inline main.js]: you chose to use line breaks here. What is your indentation rule? What do you think when do line breaks contribute to the reading flow? when do you think they disrupt the flow?

File: main.js [hunk 2]
```diff
@@ -1,60 +1,59 @@
...
+      `;
+    }
+    return response.json();
```
Comment (motine) [inline main.js]: what type of object does `json()` return?

File: main.js [hunk 3]
```diff
@@ -1,60 +1,59 @@
...
+  })
+  .then(json => {
+    const quotes = json;
```
Comment (motine) [inline main.js]: nice touch with the variable. why did you choose to do this?

File: main.js [hunk 4]
```diff
@@ -1,60 +1,59 @@
...
+    init(quotes);
+  });
+
+function init(quotes) {
+  const displayArea = document.getElementById('display-area');
+  bindEvents();
+  displayNewQuote();
```
Comment (motine) [inline main.js]: I like the breaking up into functions.

File: main.js [hunk 5]
```diff
@@ -1,60 +1,59 @@
...
+
+  function bindEvents() {
```
Comment (motine) [inline main.js]: all the subsequent functions are within the init function. is there a reason for it? do you think this is semantically correct?

File: main.js [hunk 6]
```diff
@@ -1,60 +1,59 @@
...
+    displayArea.classList.toggle('show');
+    updateTwitterIntent(quote);
+  };
+
+  function getRandomIndex(array) {
+    return Math.floor(Math.random() * array.length);
+  }
+
+  function getRandomQuote(quotes) {
+    const index = getRandomIndex(quotes);
```
Comment (motine) [inline main.js]: mmm, I like the breaking up of functions above. however, I think we are getting a little too detailed here. usually there is two reasons for using a separate function: you need to reuse it or you want to show a new semantic block. I think `getRandomIndex` and `getRandomQuote` may even be together (only my opinion). We need to ask ourselves: does the breaking up improve readability more than the extra boilerplate (function definition) harms readability.

File: main.js [hunk 7]
```diff
@@ -1,60 +1,59 @@
...
+    return quotes[index];
+  }
+
+  function handleNewQuoteClick() {
+    displayArea.classList.toggle('show');
+    setTimeout(displayNewQuote, 350);
```
Comment (motine) [inline main.js]: what happens if you hammer on the button several times super fast?

File: main.js [hunk 8]
```diff
@@ -1,60 +1,59 @@
...
+  }
+
+  function updateTwitterIntent(quote) {
+    const twitterLink = document.getElementById('tweet-quote-link');
+    const intentAddress = `https://twitter.com/intent/tweet?text=“${quote.text}”  — ${quote.movie}`;
```
Comment (motine) [inline main.js]: I am pretty sure `“` is not allowed in a URL. most browsers will be forgiving though. research: so what characters are valid in URLs? how do you escape letters that are not allowed?

File: index.html [hunk 9]
```diff
@@ -17,7 +17,7 @@
         <blockquote>
```
Comment (motine) [inline index.html:17]: nice work, Matt. the code is easy to read and well structured. please checkout my comments below.

after you have implemented the external url, maybe we could go one step further and define a [class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) for the functionality. This will not be particularly useful for this relatively small example, but it may serve as a good intro for later exercises/frameworks.

**Review by motine — CHANGES_REQUESTED** [review]

File: main.js (hunk 1 above)
Comment (MatthewBerglund) [reply main.js]: I'm not sure what you mean by "indentation rule". Could you explain?

The main reason I used line breaks was because the template literal was running off the page, but I guess you could just turn on word wrap to fix that. I think the entire `fetch().then().then()` chain is cumbersome to read, but it's definitely easier without the line breaks in the template literal.

As to when line breaks contribute to/disrupt the reading flow: I would say it comes down to how much "cognitive load" a given line of code puts on the reader, how many statements the line contains and how easy/difficult each statement is to read.

File: main.js (hunk 2 above)
Comment (MatthewBerglund) [reply main.js]: It returns a promise.

File: main.js (hunk 8 above)
Comment (MatthewBerglund) [reply main.js]: Thanks for the heads-up. 

If I understand correctly, unreserved unicode characters like `“` `”` `‘` `’` do not need to be encoded in URLs in HTML because "_browsers will take care of them per the URL parsing algorithm in the HTML spec_" (I got this from the **Update** comment from Mathias Bynens highlighted in yellow [here](https://perishablepress.com/stop-using-unsafe-characters-in-urls/#html-javascript)). So it sounds to me like `https://twitter.com/intent/tweet?text=“This is a quote”` is a valid and safe URL when it is assigned as the value of the `href` attribute on an anchor element in HTML.

This appears to be true for Chrome and Firefox, although both seem to "hide" the encoding in the address bar. I copy+pasted the URLs from the browser into a plain text editor, and this is more or less what you see: `https://twitter.com/intent/tweet?text=%E2%80%9CThis%20is%20a%20quote%E2%80%9D%`.

However, the quotes themselves do contain a number of reserved characters: `!` `?` `,` (e.g. "You talkin’ to me?", "You can't handle the truth!"). Even though the twitter intent links work, I imagine it would be a good idea to percent encode these characters anyways.

Comment (motine) [reply main.js]: yep. I agree. if in doubt make sure to encode them.

File: main.js (hunk 2 above)
Comment (motine) [reply main.js]: yep. exactly :)

File: main.js (hunk 1 above)
Comment (motine) [reply main.js]: indentation is a [whole thing](https://en.wikipedia.org/wiki/Indentation_style). some companies have style guides which outline where a bracket goes, when to use line breaks, etc.. there are so called linters which do this job for you. here [an example](https://eslint.org). there is much discussion about these "coding styles". maybe you pick one for your own projects. personally, I do not care to much which coding style one uses as long as they are applied consistently.

Comment (motine) [reply main.js]: I agree with the disrupted reading flow. this is why I asked. maybe experiment a little with the code above. Everyone has different styles, but personally I value people who think about such details quite a bit more.

Comment (motine) [reply main.js]: there is an extension to javascript's `fetch.then.then`. it is called async/await. maybe do some research about it. I found [this one](https://wesbos.com/javascript/12-advanced-flow-control/70-async-await) right away. maybe check it out. please also check caniuse.com to be aware of how well supported the feature is.

Comment (motine) [reply main.js]: I am aware that my comments seem very detailed. however I think being able to discuss your coding style or relatively new technologies such as async/await may be deem very helpful in a work context later. (I sure would be impressed by a junior who is making an effort to maximise readability.

File: main.js (hunk 3 above)
Comment (MatthewBerglund) [reply main.js]: Honestly I just thought it made it clearer that the response object `json` contains the quote data that the app uses. So, for the sake of readability.

Comment (motine) [reply main.js]: ah. understand. maybe we can name argument something like `then(quoteJson) =>`

File: main.js (hunk 5 above)
Comment (MatthewBerglund) [reply main.js]: I did this so that when the `quotes` argument is passed to `init()`, it is scoped properly for the other functions to access it. As to whether it is semantically correct -- I don't know. Much of what is going on inside of `init()` doesn't have anything to do with initializing the app, so maybe that would be an argument against taking this approach.

I suppose I could just as easily declare a global variable `quotes` at the very top and bind it to the `json` response object at line 13 without passing it to `init()`.

File: main.js (hunk 7 above)
Comment (MatthewBerglund) [reply main.js]: Hmm, it looks like it just cycles through the quotes without playing the animation properly. I think I can get around this by disabling the button at the start of the animation and enabling it again just after the new quote fades in.

File: index.html (hunk 9 above)
Comment (MatthewBerglund) [reply index.html:17]: Hey Tom,

I've got the external API working, but I'm not sure if I want to actually keep this functionality. The main problem is that the API doesn't work like it's supposed to because the quote data contains a lot of errors. Apart from the quotes containing many spelling and capitalization mistakes, the "censored" feature which allows you to fetch only quotes that do not contain "adult language" doesn't work properly because the author hasn't correctly identified which quotes actually contain adult language: the API censors quotes that do not contain adult language, and fetches ones that do.

Basically I would find it embarrassing/inappropriate to have some of those quotes show up in my app, if you know what I mean.

Should I finish the API version so that you can review it, and then change the code afterwards?

Comment (motine) [reply index.html:17]: ah that is annoying. I am sorry to hear.
let me have a look at it if you don't mind. after all we are doing this to learn. so let's use the chance.

File: main.js [hunk 10]
```diff
@@ -1,60 +1,58 @@
...
+    .then(response => {
+      if (!response.ok) {
+        const quoteBox = document.getElementById('quote-box');
+        quoteBox.textContent = `Network request failed with status ${response.status}: ${response.statusText}. Please try again.`;
+      }
+      return response.json();
+    })
+    .then(quoteJson => {
+      const quotePara = document.getElementById('quote');
+      quotePara.innerText = `“${quoteJson.quote}”`;
```
Comment (motine) [inline main.js:19]: nice use of the template strings

File: main.js [hunk 11]
```diff
@@ -1,60 +1,58 @@
...
+
+      const refCite = document.getElementById('ref');
```
Comment (motine) [inline main.js:21]: maybe let's use a little more verbose element names next time. `ref` can be anything.

File: main.js [hunk 12]
```diff
@@ -1,60 +1,58 @@
...
-  const quote = getRandomQuote(quotes);
+  fetch('https://movie-quote-api.herokuapp.com/v1/quote?censored')
+    .then(response => {
+      if (!response.ok) {
+        const quoteBox = document.getElementById('quote-box');
+        quoteBox.textContent = `Network request failed with status ${response.status}: ${response.statusText}. Please try again.`;
+      }
+      return response.json();
+    })
+    .then(quoteJson => {
```
Comment (motine) [inline main.js:17]: if you are looking for the next challenge: maybe rewrite this as [async/await](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await). if you do so, maybe let us merge this merge request first and then start a new one. i think having more split MRs makes the evolution of the code clearer.

File: main.js [hunk 13]
```diff
@@ -1,60 +1,58 @@
-const quotes = [
```
Comment (motine) [inline main.js:1]: nice work, @MatthewBerglund. Nice and readable. No comments from my side. Do you want to merge or shall I?

**Review by motine — APPROVED** [review]

File: main.js (hunk 13 above)
Comment (MatthewBerglund) [reply main.js:1]: Thanks, @motine. I think I'm going to first revert back to a version of the code that fetches `quotes.json` instead of the external API because of the inappropriate language issue I mentioned before. Then, I'll rewrite the code to use async/await instead of `fetch().then().then()` . 

Question: Should I merge first and then replace the external API? Or replace and then merge?

Comment (motine) [reply main.js:1]: whohoo, I have an idea. what if we have multiple branches (good exercise with git)... 
the master branch will be the version you want to show off. then we have one branch with the fetch-version and one with the async-version.

approach:
1. merge this merge request without fetch changes to master
2. create a new branch, directly commit your changes for fetch into this branch
3. create a new branch from the fetch branch (A), create yet another branch (B), apply the async changes to B. make a pull request from B to A.

these approaches should exercise your git-muscles quite a bit because we are doing 3 different approaches here.

pro tipp copy your repo into another folder on your harddrive if you screw something up.

Comment (MatthewBerglund) [reply main.js:1]: Okay, sounds good!
